File: clients/cpp/model/Sound.cpp
#include "Sound.hpp"

namespace model {

Sound::Sound(int typeIndex, int unitId, model::Vec2 position) : typeIndex(typeIndex), unitId(unitId), position(position) { }

// Read Sound from input stream
Sound Sound::readFrom(InputStream& stream) {
    int typeIndex = stream.readInt();
    int unitId = stream.readInt();
    model::Vec2 position = model::Vec2::readFrom(stream);
    return Sound(typeIndex, unitId, position);
}

// Write Sound to output stream
void Sound::writeTo(OutputStream& stream) const {
    stream.write(typeIndex);
    stream.write(unitId);
    position.writeTo(stream);
}

// Get string representation of Sound
std::string Sound::toString() const {
    std::stringstream ss;
    ss << "Sound { ";
    ss << "typeIndex: ";
    ss << typeIndex;
    ss << ", ";
    ss << "unitId: ";
    ss << unitId;
    ss << ", ";
    ss << "position: ";
    ss << position.toString();
    ss << " }";
    return ss.str();
}

}

File: clients/rust/src/model/sound.rs
use super::*;

/// Sound heard by one of your units
#[derive(Clone, Debug)]
pub struct Sound {
    /// Sound type index (starting with 0)
    pub type_index: i32,
    /// Id of unit that heard this sound
    pub unit_id: i32,
    /// Position where sound was heard (different from sound source position)
    pub position: model::Vec2,
}

impl trans::Trans for Sound {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.type_index.write_to(writer)?;
        self.unit_id.write_to(writer)?;
        self.position.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let type_index: i32 = trans::Trans::read_from(reader)?;
        let unit_id: i32 = trans::Trans::read_from(reader)?;
        let position: model::Vec2 = trans::Trans::read_from(reader)?;
        Ok(Self {
            type_index,
            unit_id,
            position,
        })
    }
}